File: app/src/main/java/at/shockbytes/corey/body/wearable/AndroidWearManager.java
package at.shockbytes.corey.body.wearable;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentActivity;
import android.util.Log;
import android.widget.Toast;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.wearable.MessageApi;
import com.google.android.gms.wearable.MessageEvent;
import com.google.android.gms.wearable.PutDataRequest;
import com.google.android.gms.wearable.Wearable;
import com.google.gson.Gson;

import java.util.List;

import javax.inject.Inject;

import at.shockbytes.corey.R;
import at.shockbytes.corey.common.core.workout.model.Workout;
import at.shockbytes.corey.storage.StorageManager;
import at.shockbytes.corey.workout.WorkoutManager;
import rx.functions.Action1;

/**
 * @author Martin Macheiner
 *         Date: 18.03.2017.
 */

public class AndroidWearManager implements WearableManager,
        GoogleApiClient.OnConnectionFailedListener, GoogleApiClient.ConnectionCallbacks, MessageApi.MessageListener {

    private Context context;
    private GoogleApiClient apiClient;

    private Gson gson;
    private WorkoutManager workoutManager;
    private SharedPreferences preferences;
    private StorageManager storageManager;

    @Inject
    public AndroidWearManager(Context context, WorkoutManager workoutManager,
                              StorageManager storageManager, SharedPreferences preferences,
                              Gson gson) {
        this.context = context;
        this.workoutManager = workoutManager;
        this.storageManager = storageManager;
        this.preferences = preferences;
        this.gson = gson;
    }

    @Override
    public void connect(FragmentActivity activity) {

        if (apiClient == null) {
            apiClient = new GoogleApiClient.Builder(context)
                    .addApi(Wearable.API)
                    .enableAutoManage(activity, 1, this)
                    .addConnectionCallbacks(this)
                    .addOnConnectionFailedListener(this)
                    .build();
        }
    }

    @Override
    public void synchronizeWorkouts(List<Workout> workouts) {

        PutDataRequest request = PutDataRequest.create("/workouts");
        String serializedWorkouts = gson.toJson(workouts);
        request.setData(serializedWorkouts.getBytes());

        Wearable.DataApi.putDataItem(apiClient, request);
    }

    @Override
    public void onPause() {
        Wearable.MessageApi.removeListener(apiClient, this);
        Wearable.CapabilityApi.removeLocalCapability(apiClient, context.getString(R.string.capability));
    }

    @Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {

        Toast.makeText(context,
                "Cannot connect to wearable: " + connectionResult.getErrorMessage(),
                Toast.LENGTH_LONG).show();
    }

    @Override
    public void onConnected(@Nullable Bundle bundle) {
        Wearable.MessageApi.addListener(apiClient, this);
        Wearable.CapabilityApi.addLocalCapability(apiClient, context.getString(R.string.capability));
        synchronize();
    }

    @Override
    public void onConnectionSuspended(int i) {

    }

    @Override
    public void onMessageReceived(MessageEvent messageEvent) {

        Log.wtf("Corey", "information received! + " + messageEvent.toString());
        if (messageEvent.getPath().equals("/wear_information")) {
            Toast.makeText(context, new String(messageEvent.getData()), Toast.LENGTH_SHORT).show();
            grabDataFromMessage(messageEvent);
        }

    }

    private void grabDataFromMessage(MessageEvent messageEvent) {

        String data = new String(messageEvent.getData());
        String[] s = data.split(",");

        try {

            int pulse = Integer.parseInt(s[0]);
            int workouts = Integer.parseInt(s[1]);
            int time = Integer.parseInt(s[2]);

            storageManager.updateWearWorkoutInformation(pulse, workouts, time);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void synchronize() {

        workoutManager.getWorkouts().subscribe(new Action1<List<Workout>>() {
            @Override
            public void call(List<Workout> workouts) {
                synchronizeWorkouts(workouts);
            }
        });
    }

}

File: app/src/main/java/at/shockbytes/corey/wearable/AndroidWearManager.kt
package at.shockbytes.corey.wearable

import android.content.Context
import android.net.Uri
import android.widget.Toast
import at.shockbytes.corey.R
import at.shockbytes.corey.common.core.util.WatchInfo
import at.shockbytes.corey.common.core.workout.model.Workout
import at.shockbytes.corey.data.workout.WorkoutRepository
import com.google.android.gms.wearable.CapabilityClient
import com.google.android.gms.wearable.CapabilityInfo
import com.google.android.gms.wearable.MessageClient
import com.google.android.gms.wearable.MessageEvent
import com.google.android.gms.wearable.PutDataRequest
import com.google.android.gms.wearable.Wearable
import com.google.gson.Gson

/**
 * Author:  Martin Macheiner
 * Date:    18.03.2017
 */
class AndroidWearManager(
    private val context: Context,
    private val workoutManager: WorkoutRepository,
    private val gson: Gson
) : WearableManager, MessageClient.OnMessageReceivedListener, CapabilityClient.OnCapabilityChangedListener {

    private var nodeListener: ((WatchInfo) -> Unit)? = null

    override fun onStart(nodeListener: ((WatchInfo) -> Unit)?) {
        this.nodeListener = nodeListener

        Wearable.getMessageClient(context).addListener(this)
        Wearable.getCapabilityClient(context).addLocalCapability(context.getString(R.string.capability_device))
        Wearable.getCapabilityClient(context)
                .addListener(this, Uri.parse("wear://"), CapabilityClient.FILTER_REACHABLE)
        synchronize()
    }

    override fun onPause() {
        Wearable.getMessageClient(context).removeListener(this)
        Wearable.getCapabilityClient(context).removeLocalCapability(context.getString(R.string.capability_device))
        Wearable.getCapabilityClient(context).removeListener(this)
    }

    override fun synchronizeWorkouts(workouts: List<Workout>) {

        val request = PutDataRequest.create("/workouts")
        val serializedWorkouts = gson.toJson(workouts)
        request.data = serializedWorkouts.toByteArray()

        Wearable.getDataClient(context).putDataItem(request).addOnCompleteListener { }
    }

    override fun onMessageReceived(messageEvent: MessageEvent) {

        if (messageEvent.path == "/wear_information") {
            Toast.makeText(context, String(messageEvent.data), Toast.LENGTH_SHORT).show()
            grabDataFromMessage(messageEvent)
        }
    }

    override fun onCapabilityChanged(info: CapabilityInfo) {
        if (info.nodes.size > 0) {
            val connectedNode = info.nodes.iterator().next() // Assume first node is watch
            nodeListener?.invoke(WatchInfo(connectedNode?.displayName, true))
        } else {
            nodeListener?.invoke(WatchInfo(null, false))
        }
    }

    private fun grabDataFromMessage(messageEvent: MessageEvent) {

        val s = String(messageEvent.data).split(",".toRegex()).dropLastWhile { it.isEmpty() }

        val pulse = s[0].toIntOrNull()
        val workouts = s[1].toIntOrNull()
        val time = s[2].toIntOrNull()

        if (pulse != null && workouts != null && time != null) {
            workoutManager.updateWearWorkoutInformation(pulse, workouts, time)
        }
    }

    private fun synchronize() {
        workoutManager.workouts.subscribe { workouts -> synchronizeWorkouts(workouts) }
    }
}
